### PR TITLE
Refactor source [ch124583]

### PR DESCRIPTION
**Pull request by Clebal**
- Refactored data & source structure in template-sample-app.
- Refactored hygen layer to fit new structure.
- Added hygen source to generate sources.
- Refactored data & source structure in template-skeleton.
- Added documentation.

From [this](https://app.clubhouse.io/cartoteam/story/124583/refactor-source) CH ticket.

Proposal: #164.

### CODE REVIEW

File: hygen/_templates/layer/new/prompt.js
Comment (albertoarana) [inline hygen/_templates/layer/new/prompt.js:58]: Improved:
```suggestion
  answers.type_source = 'sql'
  
  if (res) {
    const { groups: { type: sourceType } } = res
    if (SOURCE_TYPES.indexOf(sourceType) === -1) {
      throw new Error('The source has an unknown type.');
    }
    answers.type_source = sourceType
  }
```

File: hygen/_templates/layer/new/prompt.js
Comment (albertoarana) [inline hygen/_templates/layer/new/prompt.js:101]: Improved:
```suggestion
    answers.view_path = VIEWS_DIR
    
    if (answers.view.includes('views/')) {
      const selectedViewInitialPath = answers.view.split('(')[1].replace(')', '');
      answers.view_path = viewFiles.find(viewFile => {
        return viewFile.path === selectedViewInitialPath.replace('views', VIEWS_DIR)
      }).path
    }
```

File: hygen/_templates/promptUtils.js
Comment (albertoarana) [inline hygen/_templates/promptUtils.js:27]: The map maybe be a reducer and the filter function is included.

File: hygen/_templates/source/new/prompt.js
Comment (albertoarana) [inline hygen/_templates/source/new/prompt.js:37]: If source already exists, the code is executed but nothing happens because hygen has the `skip` flag if that source exists.

File: hygen/_templates/source/new/source.ejs.t
Comment (albertoarana) [inline hygen/_templates/source/new/source.ejs.t:4]: The source suffix is mandatory in input

**Review by albertoarana — CHANGES_REQUESTED** [review]

File: hygen/_templates/layer/new/inject_view_useEffect.ejs.t
Comment (Clebal) [reply hygen/_templates/layer/new/inject_view_useEffect.ejs.t:8]: @albertoarana I've added this: `answers.name = answers.name.replace('Layer', '').replace('layer', '');`. I think it's easier to allow user to optionally avoid write layer or source (in source generator). However, in the new task about layer generator, I've added that `layer` word is automatically added to name to avoid to add it in .ejs.t. If you allow user to write anything, it's going to be a mess. It's better to add a standard suffix in my opinion.

File: hygen/_templates/layer/new/prompt.js
Comment (Clebal) [reply hygen/_templates/layer/new/prompt.js:101]: Chapó 👌

File: hygen/_templates/source/new/prompt.js
Comment (Clebal) [reply hygen/_templates/source/new/prompt.js]: Fixed.

File: hygen/_templates/promptUtils.js
Comment (Clebal) [reply hygen/_templates/promptUtils.js]: @albertoarana fixed.

**Review by albertoarana — APPROVED** [review]

File: template-sample-app/template/src/components/layers/StoresLayer.js
Comment (VictorVelarde) [inline template-sample-app/template/src/components/layers/StoresLayer.js:31]: I think id should be extracted from LayerStyle object, as it is conceptually not included inside style props, what do you think?

File: template-sample-app/template/src/components/views/Kpi.js
Comment (VictorVelarde) [inline template-sample-app/template/src/components/views/Kpi.js:22]: After the good refactor you have done on Source, removing the const here and exposing it from the 'source' file, I think we should probably do the same for layers: exporting a const with the 'id' from each Layer file, and consuming it here with kpiLayer.id. That way, there is just one place for the LAYER_ID, and that is the layer itself, which IMO it makes sense...

what do you think?

File: template-sample-app/template/src/components/views/Kpi.js
Comment (VictorVelarde) [inline template-sample-app/template/src/components/views/Kpi.js:80]: this looks nice!

File: template-sample-app/template/src/components/views/datasets/Datasets.js
Comment (VictorVelarde) [inline template-sample-app/template/src/components/views/datasets/Datasets.js:61]: We're gonna remove this file (all related to that tab), but I'll wait till this gets merged, to avoid any conflict

File: template-sample-app/template/src/components/layers/StoresLayer.js
Comment (VictorVelarde) [reply template-sample-app/template/src/components/layers/StoresLayer.js:31]: An exported const at the top could be useful, as with sources

File: template-sample-app/template/src/components/views/Kpi.js
Comment (Clebal) [reply template-sample-app/template/src/components/views/Kpi.js:22]: It's being done in #167.

**Comment by VictorVelarde** [comment]
Everything looks really nice. Great work. I have tested it locally and it works ✨ 

Please have a look at the conflicts and also add a small note to CHANGELOG with the new feature.

The only strong suggestion I would do is the name of the exported function in layers. If I use MyLayer, the file is MyLayer.js, but the exported function is just `My`. I think it will be more appropiated with `export default function MyLayer()`

File: template-sample-app/template/src/components/layers/StoresLayer.js
Comment (albertoarana) [reply template-sample-app/template/src/components/layers/StoresLayer.js:31]: @VictorVelarde that change is contemplated in the next PR where @Clebal modifies the generator Layer

**Comment by albertoarana** [comment]
@VictorVelarde I like suggestion. We are going to change it

**Comment by albertoarana** [comment]
Done. I have needed to change the file names to camelCase.

File: template-sample-app/template/src/data/sources/kpiSource.js
Comment (alasarr) [inline template-sample-app/template/src/data/sources/kpiSource.js:1]: I'd simplify this to only export the default.

It adds a data structure that it's not required and it adds complexity in several parts of the code.

KPI_SOURCE_ID is available as default.id
KPI_SOURCE_COLUMNS is like have two ways to define the source at the SQL and the const. 

It's a key concept for new users and we need to simplify this as much as possible. 

If someone wants to complicate more to have a more robust architecture is fine, but not for all the users

**Review by alasarr — CHANGES_REQUESTED** [review]

**Review by VictorVelarde — APPROVED** [review]
Fixed some conflicts after merging client-side PR

**Pull request merged**
